### PR TITLE
build: introduce release-series.txt

### LINKED ISSUE #91306: build: use a checked-in file to determine binary version instead of git tags
**Describe the problem**

Currently, the binary version is set during build time and determined using `git describe --tags` [1][2][3]. Changing the binary version is done by tagging master and consequently 1) without any corresponding PR and 2) without running our test suite. This means tagging master can have the undesired consequence of breaking tests on master (e.g., version upgrade tests, that now start running new migrations), which is very disruptive.

Ideally, we want these updates to happen by changing a file that is checked in, so that any test failures are caught before the change is merged in the PR that updates the version file.

**Roachtest**

`roachtest` also independently looks at git tags [4]. It should be reading from the same checked-in version file.

**Related issues**

* #75287: missing tags can also lead to wrong binary versions being reported in telemetry. Not relying on git tags would fix that.

[1] https://github.com/cockroachdb/cockroach/blob/master/pkg/build/BUILD.bazel#L21
[2] https://github.com/cockroachdb/cockroach/blob/master/build/bazelutil/stamp.sh#L47
[3] https://github.com/cockroachdb/cockroach/blob/master/pkg/build/info.go#L42
[4] https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/test_registry.go#L202

Jira issue: CRDB-21197

### PR DESCRIPTION
**Pull request by renatolabs**
This commit adds the `pkg/build/release-series.txt` file, intended to
contain the current release series (i.e., MAJOR.MINOR). Previously,
the `build` package would obtain this information by parsing the
build tag from git. The main advantages of this change are:

1. No reliance on git tags being available on the machine used to
build `cockroach`, thus reducing the chances of accidently
misreporting a binary's version (see #75287).
2. Changing of the release series requires a pull request and, as a
consequence, we are able to verify that tests don't break because of
that. Relying on tags alone may expose bugs that can lead to broken
bors builds.

Note that release builds still rely on the git tag when reporting the
binary version (which is only used in a few places at the
moment). This is for readability, so that log entries and collected
metrics use a binary version closer to what a user expects.

Resolves #91306.

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
Notes for reviewers:

* I expect tests to break; I pushed this sooner to get some feedback on the direction of the changes.
* I considered the advantages and disadvantages of having a “full” version file (including patch version) versus just the release series (as is done in this PR). The former would involve some changes in our release process and automation, as a lot of it relies on git tags. Having just the release series makes it much easier to integrate with the existing runbook.

One of the main disadvantages of the current approach is that `build.BinaryVersion()` no longer returns a version parseable by the `version` package. This is currently not a big problem, as I don’t think any of the callers of that function actually try to parse the version.

I’d also like to hear if the new logic in `BinaryVersion()` to still use tags in released binaries makes sense.

**Review by rickystewart — CHANGES_REQUESTED** [review]
The overall change seems roughly fine, but note that this does NOT "resolve #91306", nor does it have any particular effect on #75287. We are still using `git` tags during the build and embedding them into the binary. The description of the PR/commit message is incorrect in that respect. If that is what you want to accomplish with this change, then let me know and we can go over the details of the changes that would be necessary. There is more work to do in the build system and in the various scripts we have setup for release anyway. Otherwise, at the very least the PR description should be updated so it is accurate.

**Comment by rickystewart** [comment]
If it's up to me I prefer to have the entire version in a file rather than just the release series. This might complicate the runbooks some but it would make the actual release process simpler (and presumably the extra stuff can be removed from the runbooks after 22.2 is decommissioned), as well as allowing us to actually solve the problems that you appear to be aiming to solve given the PR description.

**Comment by renatolabs** [comment]
> note that this does NOT "resolve https://github.com/cockroachdb/cockroach/issues/91306"

Fair enough! Git tags were being used for two reasons:

* As another data point about the binary (displayed in `cockroach version`, for example)
* Used to infer the binary version (majority of cases)

My main concern was removing the second use as it relates to tests that rely on the binary version (and avoid broken bors builds after a new tag is added), not to completely eliminate the use of tags from the release process. As I noted in the description, this PR only includes a release-series file and as a consequence I opted to leave the logic unchanged for release binaries.

That said, if most people think we should work toward a world where git tags are completely eliminated as a means to determine binary version, I'm happy to work on that as well.

**Comment by rickystewart** [comment]
> As I noted in the description, this PR only includes a release-series file and as a consequence I opted to leave the logic unchanged for release binaries.

The logic is unchanged not just for release binaries, but ALL binaries. The build process for development binaries, custom builds, etc. also all uses `git tag`s. This PR does not change any of the structural issues we've discussed here or in the above issues.

> That said, if most people think we should work toward a world where git tags are completely eliminated as a means to determine binary version, I'm happy to work on that as well.

That is certainly my preference, but I'm not really a stakeholder here and I'm happy to leave it up to the rest of the group.

**Comment by renatolabs** [comment]
> The build process for development binaries, custom builds, etc. also all uses git tags. This PR does not change any of the structural issues we've discussed here or in the above bugs.

Let me rephrase that then: the goal here was for `build.BinaryVersion*()` to not depend on git tags _for development builds_ so that behavior is the same during tests regardless of git tags at build time. Tags are still a part of the build process and embedded in the binary (as you said). But tests won't fail after a tag is added and also engineers running `roachtest` will not have to remember to fetch tags before running a test.

> That is certainly my preference, but I'm not really a stakeholder here and I'm happy to leave it up to the rest of the group.

Thanks, this is good to know! I'm also curious to hear opinions from more folks involved in the release process (pinging @celiala, @rail, @jlinder -- just because you guys are part of the `release-eng-prs` group 🙂)

**Comment by dt** [comment]
When I suggested this, I think I was thinking we wouldn't mess with what `BinaryVersion()` returns, but would instead where we currently call it and strip off all but the release series -- 23.1 -- to look in the map to find its predecessor for testing, there we would depend on this new "ReleaseSeries()` thing instead. That feels a little cleaner to me than having `BinaryVersion` be a somewhat-git-somewhat-not sprintf? just leave it as fully up to git but then don't care about its answers in how we run tests?

**Comment by renatolabs** [comment]
> when this change lands, are we updating our release processes to make sure that the `release-series.txt` file is updated?

We're still discussing the direction we're going to go here; but yes, regardless of what we end up doing, I plan to update our release runbooks accordingly.

**Comment by renatolabs** [comment]
> I think I was thinking we wouldn't mess with what BinaryVersion() returns

That's also an option; however, one downside is that we could break bors builds as a result of tagging if some test logic inadvertently relies on the string returned by `BinaryVersion` (as happened in #91287, for example)

**Comment by dt** [comment]
> if some test logic inadvertently relies on the string returned by BinaryVersion

well, yeah, but I’m saying we don’t rely the string returned by BinaryVersion (ie put a comment on it being like ‘if you need to know what release this is, use ReleaseSeries()” etc)

**Comment by celiala** [comment]
> > That is certainly my preference, but I'm not really a stakeholder here and I'm happy to leave it up to the rest of the group.
> 
> Thanks, this is good to know! I'm also curious to hear opinions from more folks involved in the release process (pinging @celiala, @rail, @jlinder -- just because you guys are part of the release-eng-prs group 🙂)

I haven't yet had a chance to fully digest the discussion on this PR (or the code itself), but I wanted to quickly chime in that I'm a +1 for a world where we don’t rely on git tags to determine binary version.

**Comment by rail** [comment]
I like the idea! There was a [ticket](https://cockroachlabs.atlassian.net/browse/RE-107), very similar to this one. I wonder if we should combine them together or gradually move...

**Comment by renatolabs** [comment]
Thank you for everyone's inputs! I'll close this one in favour of a future PR.
